### PR TITLE
Generate "Decode" for "enums" that will directly decode

### PR DESCRIPTION
**Pull request by witriew**
Add templating for generating code for "enums" that will utilize the new
`stream.Reader`, performing the same duties as `binary.Decode` and `FromWire`,
but without going through an intermediary format.

### CODE REVIEW

File: gen/enum.go [hunk 1]
```diff
@@ -190,6 +191,32 @@ func enum(g Generator, spec *compile.EnumSpec) error {
...
+		//     return <$enumName>(0), err
+		//   }
+		//   return <$v>, nil
+		func (<$v> *<$enumName>) Decode(<$sr> <$stream>.Reader) error {
+			<- $rVal := newVar "rVal" ->
+			<$rVal>, err := <$sr>.ReadInt32()
+			if err != nil {
+				return err
+			}
+
```
Comment (abhinav) [inline gen/enum.go:215]: small nit: delete this empty line please.
```suggestion
```

File: gen/roundtrip_test.go [hunk 2]
```diff
@@ -81,3 +82,47 @@ func assertBinaryRoundTrip(t *testing.T, w wire.Value, message string) (wire.Val
 
 	return newV, true
 }
+
+// assertStreamingRoundTrip checks that the streaming Encode/Decode returns the same
+// value.
+//
+// Note: Leverages 'ToWire' and regular binary.Encode until a 'stream.Writer'
+// implementation is plumbed through.
+func assertStreamingRoundTrip(t *testing.T, x thriftType, v wire.Value, msg string, args ...interface{}) bool {
```
Comment (abhinav) [inline gen/roundtrip_test.go:91]: So, this is probably fine for now, but possibly once we have the writer, an alternative is that this is the new assertRoundTrip, and does a cross-product of streaming/non-streaming encode/decode.

Like,

```
streams := []struct{ encode, decode bool }{
  {false, false},
  {false, true},
  {true, false},
  {true, true}
}

for _, stream := range streams {
  _, canStream := x.(streamingThriftType)
  
  // Skip cases where we can't stream and streaming is requested.
  if !canStream && (stream.encode || stream.decode) {
    continue
  }
  
  name := ...
  t.Run(name, func(t *testing.T) {
    var buff bytes.Buffer
    if stream.encode {
        x.Encode(binary.Writer(&buff))
    } else {
        binary.Encode(x.ToWire(), &buff)
    }
    
    if stream.decode {
        // ...
    } else {
        // ...
    }
  })
}
```

Comment (witriew) [reply gen/roundtrip_test.go:91]: Yep, I was imagining this as well.

Comment (witriew) [reply gen/roundtrip_test.go:92]: Added a comment to reflect that we want to do that eventually.

File: gen/enum.go [hunk 3]
```diff
@@ -190,6 +191,32 @@ func enum(g Generator, spec *compile.EnumSpec) error {
 			return nil
 		}
 
+		<$sr := newVar "sr">
+		// Decode reads off the encoded <$enumName> directly off of the wire.
+		//
+		//   sp, err := binaryProtocol.(stream.Protocol)
+		//   if err != nil {
+		//     return <$enumName>(0), err
+		//   }
```
[the remark this answers is not in the document]
Comment (witriew) [reply gen/enum.go]: Cleaned this up a bit, lemme know how it looks.

File: gen/internal/tests/enum_conflict/enum_conflict.go [hunk 4]
```diff
@@ -122,6 +123,30 @@ func (v *RecordType) FromWire(w wire.Value) error {
 	return nil
 }
 
+// Decode reads off the encoded RecordType directly off of the wire.
+//
+//   sp, err := binaryProtocol.(stream.Protocol)
+//   if err != nil {
+//     return RecordType(0), err
+//   }
```
Comment (abhinav) [inline gen/internal/tests/enum_conflict/enum_conflict.go:131]: outdated generated code

**Review by abhinav — COMMENTED** [review]
Generated code is out of date, and needs to be rebased, but lgtm otherwise.

File: gen/enum.go [hunk 5]
```diff
@@ -190,6 +191,26 @@ func enum(g Generator, spec *compile.EnumSpec) error {
...
+		//
+		//   sReader := BinaryStreamer.Reader(reader)
+		//
+		//   var <$v> <$enumName>
+		//   if err := <$v>.Decode(sReader); err != nil {
+		//     return <$enumName>(0), err
+		//   }
+		//   return <$v>, nil
+		func (<$v> *<$enumName>) Decode(<$sr> <$stream>.Reader) error {
+			<- $rVal := newVar "rVal" ->
```
Comment (abhinav) [inline gen/enum.go:205]: optional nitty nit: I would have just named this "x", "i", or "val". There's not much ambiguity here; it's a really small function to require the "rVal" disambiguation.

**Review by abhinav — APPROVED** [review]
LGTM

File: gen/util_for_test.go [hunk 6]
```diff
@@ -37,3 +38,15 @@ type thriftType interface {
 	ToWire() (wire.Value, error)
 	FromWire(wire.Value) error
 }
+
+// streamingThriftType is implemented by all generated types that know how to
+// write and read themselves to the Thrift Protocol, skipping over the
+// intermediary wire.Type
+//
+// Only requires Decode until Encode is plubmed through to a stream.Writer.
```
Comment (mh-park) [inline gen/util_for_test.go:46]: spelling nit:
```suggestion
// Only requires Decode until Encode is plumbed through to a stream.Writer.
```

**Review by mh-park — APPROVED** [review]
LGTM

**Pull request merged**